File: domain/src/test/java/mixter/domain/identity/FakeSessionProjectionRepository.java
package mixter.domain.identity;

import org.assertj.core.util.Sets;

import java.util.Optional;
import java.util.Set;

public class FakeSessionProjectionRepository implements SessionProjectionRepository {

    private Set<SessionProjection> sessions = Sets.newHashSet();

    public Set<SessionProjection> getSessions() {
        return sessions;
    }

    @Override
    public void save(SessionProjection sessionProjection) {
        sessions.add(sessionProjection);
    }

    @Override
    public void replaceBy(SessionProjection sessionProjection) {
        sessions.removeIf(s -> s.getSessionId().equals(sessionProjection.getSessionId()));
        sessions.add(sessionProjection);
    }

    @Override
    public Optional<SessionProjection> getById(SessionId sessionId) {
        return Optional.empty();
    }
}

File: domain/src/test/scala/mixter/domain/identity/FakeSessionProjectionRepository.scala
package mixter.domain.identity

import org.scalatest.{Suite, SuiteMixin}


trait WithSessionRepository extends SuiteMixin { this: Suite =>
  def withSessionRepository(test: FakeSessionProjectionRepository=>Any): Any = {
    val sessionRepository = new FakeSessionProjectionRepository()
    test(sessionRepository)
  }


}

private[identity] class FakeSessionProjectionRepository() extends SessionProjectionRepository {
  var sessions = Map.empty[SessionId,SessionProjection]

  def getSessions:Set[SessionProjection] =
    sessions.values.toSet

  override def save(sessionProjection: SessionProjection): Unit =
    sessions+=sessionProjection.sessionId -> sessionProjection

  override def replaceBy(sessionProjection: SessionProjection): Unit =
    save(sessionProjection)

  override def getById(id: SessionId): Option[SessionProjection] = sessions.get(id)
}
